src/kernel/state/makeVatState.js

```diff
@@ -0,0 +1,213 @@
+  const state = {
+    kernelSlotToVatSlot: new Map(),
+    vatSlotToKernelSlot: new Map(),
+
+    // make these IDs start at different values to detect errors
+    // better
+    nextIDs: {
+      import: 10,
+      promise: 20,
+      resolver: 30,
+      deviceImport: 40,
+    },
+    transcript: [],
+  };
+
+  function createKeyFromObj(obj) {
+    const keyItems = [];
+    Object.getOwnPropertyNames(obj).forEach(k => keyItems.push(k, obj[k]));
+    return keyItems.join('-');
+  }
+
+  const allowedVatSlotTypes = [
+    'export',
+    'import',
+    'deviceImport',
+    'promise',
+    'resolver',
+  ];
+
+  const allowedKernelSlotTypes = [
+    'export',
+    'import',
+    'device',
+    'promise',
+    'resolver',
+  ];
+
+  function checkVatSlot(slot) {
+    const properties = Object.getOwnPropertyNames(slot);
+    insist(
+      properties.length === 2,
+      `wrong number of properties for a vatSlot ${JSON.stringify(slot)}`,
+    );
+    Nat(slot.id);
+    insist(
+      allowedVatSlotTypes.includes(slot.type),
+      `unknown slot.type in '${JSON.stringify(slot)}'`,
+    );
+  }
+
+  function checkKernelSlot(slot) {
+    const properties = Object.getOwnPropertyNames(slot);
+    insist(
+      properties.length === 3 ||
+        (properties.length === 2 &&
+          (slot.type === 'promise' || slot.type === 'resolver')),
+      `wrong number of properties for a kernelSlot ${JSON.stringify(slot)}`,
+    ); // a kernel slot has a vatID property, unless it is a promise or a resolver
+    Nat(slot.id);
+    insist(
+      allowedKernelSlotTypes.includes(slot.type),
+      `unknown slot.type in '${JSON.stringify(slot)}'`,
+    );
+  }
+
+  function createKernelSlotKey(kernelSlot) {
+    checkKernelSlot(kernelSlot);
+    return createKeyFromObj(kernelSlot);
+  }
+
+  function createValSlotKey(vatSlot) {
+    checkVatSlot(vatSlot);
+    return createKeyFromObj(vatSlot);
+  }
+
+  function getNextVatSlotID(type) {
+    insist(
+      allowedVatSlotTypes.includes(type),
+      `type ${type} is not an allowed vat slot type`,
+    );
+    return state.nextIDs[type];
+  }
+
+  function setNextVatSlotID(type, newNextID) {
+    insist(
+      allowedVatSlotTypes.includes(type),
+      `type ${type} is not an allowed vat slot type`,
+    );
+    state.nextIDs[type] = newNextID;
+  }
+
+  function getVatSlotTypeFromKernelSlot(kernelSlot) {
+    switch (kernelSlot.type) {
+      case 'export':
+        return 'import';
+      case 'device':
+        return 'deviceImport';
+      case 'promise':
+        return 'promise';
+      case 'resolver':
+        return 'resolver';
+      default:
+        throw new Error('unrecognized kernelSlot type');
+    }
+  }
+
+  function allocateNextVatSlotID(kernelSlot) {
+    const vatSlotType = getVatSlotTypeFromKernelSlot(kernelSlot);
+    const i = getNextVatSlotID(vatSlotType);
+    setNextVatSlotID(vatSlotType, i + 1);
+    return i;
+  }
+
+  function mapVatSlotToKernelSlot(vatSlot, throwIfNotFound) {
```

Comment:
the `throwIfNotFound` parameter should go away and this should just always throw if the entry wasn't already present. The two cases in `vatManager` didn't do this check were mistakes: it should always throw rather than return `undefined`. Vats can only reference things that were given to them, or presences they're newly exporting.